Fix sign-compare violations in torch_python

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75085
* #75084
* #75082
* __->__ #75079
* #75083
* #75080
* #75081

Prerequisite change for enabling `-Werror=sign-compare` across PyTorch repo

malfet committed:
Fix sign-compare violations in torch_python

Prerequisite change for enabling `-Werror=sign-compare` across PyTorch repo

[ghstack-poisoned]

albanD commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -937,7 +937,7 @@ bool FunctionSignature::parse(PyObject* self, PyObject* args, PyObject* kwargs,
   }
 
-  if (nargs > max_pos_args && !allow_varargs_intlist) {
+  if (nargs > static_cast<decltype(nargs)>(max_pos_args) && !allow_varargs_intlist) {
```
Can we get `nargs` about into a `size_t` type above?

albanD commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -1093,7 +1093,7 @@ PythonArgs PythonArgParser::raw_parse(PyObject* self, PyObject* args, PyObject*
 void PythonArgParser::print_error(PyObject* self, PyObject* args, PyObject* kwargs, PyObject* parsed_args[]) {  // NOLINT
   // NOLINTNEXTLINE(clang-analyzer-core.NullDereference)
-  auto num_args = PyTuple_GET_SIZE(args) + (kwargs ? PyDict_Size(kwargs) : 0);
+  auto num_args = static_cast<size_t>(PyTuple_GET_SIZE(args) + (kwargs ? PyDict_Size(kwargs) : 0));
```
Is changing the auto with size_t a way to get the same behavior?
Or the static_cast is preferred here?

albanD commented on torch/csrc/utils/tensor_numpy.cpp:
```diff
@@ -371,7 +371,7 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
     PyObject *py_strides = PyDict_GetItemString(cuda_dict, "strides");
     if (py_strides != nullptr && py_strides != Py_None) {
-      if (PySequence_Length(py_strides) == -1 || PySequence_Length(py_strides) != sizes.size()) {
+      if (PySequence_Length(py_strides) == -1 || PySequence_Length(py_strides) != static_cast<Py_ssize_t>(sizes.size())) {
```
Don't we prefer to cast the result of `PySequence_Length` to `size_t`?
From a quick search, looks like `Py_ssize_t` is a `int` only.

malfet committed:
Update on "Fix sign-compare violations in torch_python"


Prerequisite change for enabling `-Werror=sign-compare` across PyTorch repo

[ghstack-poisoned]

albanD reviewed (approved):
Thanks for the update!
